cockroachdb/templates/job.init.yaml

```diff
@@ -86,29 +88,140 @@ spec:
-              {{- end }}
-              --host={{ template "cockroachdb.fullname" . }}-0.{{ template "cockroachdb.fullname" . -}}
-                     :{{ .Values.service.ports.grpc.internal.port | int64 }}
-              2>&1);
-              initRC="$?";
-              echo $initOUT;
-              [[ "$initRC" == "0" ]] && exit 0;
-              [[ "$initOUT" == *"cluster has already been initialized"* ]] && exit 0;
-              sleep 5;
-              done
+          - /bin/bash
+          - -c
+          - >-
+            {{- if $isClusterInitEnabled }}
+              initCluster() {
+                while true; do
+                  local output=$(
+                    set -x;
+
+                    /cockroach/cockroach init \
+                      {{- if .Values.tls.enabled }}
+                      --certs-dir=/cockroach-certs/ \
+                      {{- else }}
+                      --insecure \
+                      {{- end }}
+                      {{- with index .Values.conf "cluster-name" }}
+                      --cluster-name={{.}} \
+                      {{- end }}
+                      --host={{ template "cockroachdb.fullname" . }}-0.{{ template "cockroachdb.fullname" . -}}
+                              :{{ .Values.service.ports.grpc.internal.port | int64 }}
+                  2>&1);
+
+                  local exitCode="$?";
+                  echo $output;
+
+                  if [[ "$exitCode" == "0" || "$output" == *"cluster has already been initialized"* ]]
+                    then break;
+                  fi
+
+                  sleep 5;
+                done
+              }
+
+              initCluster;
+            {{- end }}
+
+            {{- if $isDatabaseProvisioningEnabled }}
+              provisionCluster() {
+                while true; do
+                  /cockroach/cockroach sql \
+                    {{- if .Values.tls.enabled }}
+                    --certs-dir=/cockroach-certs/ \
+                    {{- else }}
+                    --insecure \
+                    {{- end }}
+                    {{- with index .Values.conf "cluster-name" }}
+                    --cluster-name={{.}} \
+                    {{- end }}
+                    --host={{ template "cockroachdb.fullname" . }}-0.{{ template "cockroachdb.fullname" . -}}
+                            :{{ .Values.service.ports.grpc.internal.port | int64 }} \
+                    --execute="
+                      {{- range $clusterSetting, $clusterSettingValue := .Values.init.provisioning.clusterSettings }}
+                        SET CLUSTER SETTING {{ $clusterSetting }} = '${{ $clusterSetting | replace "." "_" }}_CLUSTER_SETTING';
+                      {{- end }}
+
+                      {{- range $user := .Values.init.provisioning.users }}
+                        CREATE USER IF NOT EXISTS {{ $user.name }} WITH
+                        {{- if $user.password }}
+                          PASSWORD '${{ $user.name }}_PASSWORD'
```

Comment:
Can you please add more datails? Can you share username(or same structured username)?

Reply:
Would be asome if you can share your values.yaml section

Reply:
there you go..
diff --git a/cockroachdb/values.yaml b/cockroachdb/values.yaml
index 9783821..05f64c1 100644
--- a/cockroachdb/values.yaml
+++ b/cockroachdb/values.yaml
@@ -1,9 +1,9 @@
 image:
-  tag: v21.1.7
+  tag: 21.1.8
   pullPolicy: IfNotPresent
-  credentials: {}
-    # registry: docker.io
+  credentials:
+    registry: docker.io
     # username: john_doe
     # password: changeme
 
@@ -336,7 +336,7 @@ storage:
   persistentVolume:
     enabled: true
 
-    size: 100Gi
+    size: 5Gi
 
     # If defined, then `storageClassName: <storageClass>`.
     # If set to "-", then `storageClassName: ""`, which disables dynamic
@@ -344,7 +344,7 @@ storage:
     # If undefined or empty (default), then no `storageClassName` spec is set,
     # so the default provisioner will be chosen (gp2 on AWS, standard on
     # GKE, AWS & OpenStack).
-    storageClass: ""
+    storageClass: "longhorn"
 
     # Additional labels to apply to the created PersistentVolumeClaims.
     labels: {}
@@ -388,21 +388,34 @@ init:
     #   memory: "128Mi"
 
   provisioning:
-    enabled: false
+    enabled: true
     # https://www.cockroachlabs.com/docs/stable/cluster-settings.html
-    clusterSettings:
+    clusterSettings: {}
       # cluster.organization: "'FooCorp - Local Testing'"
       # enterprise.license: "'xxxxx'"
-    users: []
-    # - name:
-    #   password:
+    users:
+    - name: alboroto
+      password: '!pass123'
+      options: [LOGIN, CREATEROLE]
+    - name: mlflow
+      password: mlflow
+      options: [LOGIN]
+    - name: superset
+      password: superset
+      options: [LOGIN]
     #   # https://www.cockroachlabs.com/docs/stable/create-user.html#parameters
     #   options: [LOGIN]
-    databases: []
-    # - name:
+    databases:
+    - name: mlflow
     #   # https://www.cockroachlabs.com/docs/stable/create-database.html#parameters
     #   options: [encoding='utf-8']
-    #   owners: []
+      owners: [mlflow]
+    - name: testdb
+      owners: [alboroto]
+    - name: testdb
+      owners: [alboroto]
+    - name: superset
+      owners: [superset]
     #   # Backup schedules are not idemponent for now and will fail on next run
     #   # https://github.com/cockroachdb/cockroach/issues/57892
     #   backup:
 networkPolicy:
   enabled: false

Reply:
Seems weird)

![image](https://user-images.githubusercontent.com/46970457/132985315-61d4d6d8-48de-4642-aac2-2742b788e2ac.png)

Does all your users without password or one starting from `!`?

Reply:
I will test real deployment with this values

Reply:
![image](https://user-images.githubusercontent.com/46970457/132985473-0a0722c5-ad45-4e93-8e0e-c3d3dc1d6c96.png)

![image](https://user-images.githubusercontent.com/46970457/132985548-0cc66b7c-d3ba-4083-85bb-806087c2bea3.png)

Missed namespace and quotation. It will work only in default namespace

Reply:
```
++ /cockroach/cockroach init --certs-dir=/cockroach-certs/ --host=test-release-cockroachdb-0.test-release-cockroachdb:26257
warning: node not ready to perform cluster initialization: initial connection heartbeat failed: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 10.42.2.103:26257: connect: connection refused" (retrying)
Cluster successfully initialized
GRANT

          CREATE USER IF NOT EXISTS alboroto WITH
            PASSWORD '!pass123'
            LOGIN CREATEROLE
          ;
          CREATE USER IF NOT EXISTS mlflow WITH
            PASSWORD 'mlflow'
            LOGIN
          ;
          CREATE USER IF NOT EXISTS superset WITH
            PASSWORD 'superset'
            LOGIN
          ;
          CREATE DATABASE IF NOT EXISTS mlflow
          ;
          GRANT ALL ON DATABASE mlflow TO mlflow;
          CREATE DATABASE IF NOT EXISTS testdb
          ;
          GRANT ALL ON DATABASE testdb TO alboroto;
          CREATE DATABASE IF NOT EXISTS testdb
          ;
          GRANT ALL ON DATABASE testdb TO alboroto;
          CREATE DATABASE IF NOT EXISTS superset
          ;
          GRANT ALL ON DATABASE superset TO superset;
      
Provisioning completed successfully
```

Reply:
Couldn't login with any.. I guess that special characters breaks the user creation loop.. Btw is there a way to set root password or guess it afterwards accessing to nodes?

Reply:
@alborotogarcia Does `$({{ $user.name }}_PASSWORD)` gives you correct passwords? Did you changed something else? Secret was in default namespace, so you can't mount its values to init job from `another-namespace`

Reply:
Does init job logs something?

Reply:
Also you can specify passwords only for secure clusters. Which are you running?

Reply:
You can't change root(and any) password, only create new user for now. Because ALTER USER is not supported

https://www.cockroachlabs.com/docs/stable/authorization.html#root-user

Reply:
Yes you were right, previously I created them as sealed secrets to the namespace, as I read from this PR that it should be taken from vault secrets, and as it was giving me errors I left them..
```apiVersion: bitnami.com/v1alpha1
kind: SealedSecret
metadata:
  creationTimestamp: null
  name: crdb-cockroachdb-init
  namespace: storage
spec:
  encryptedData:
    alboroto-password: AgB2TYo2b3Jd2Wwah31vxnOMoVrv6fZeAmlaEjWOyScio72YV0teLdYNaHoxDQc2h96kDtBbsFAtlPcGsv5IRWX49ebkSHaAkIKn+RtEANbhgiMnujImb7Vnxmw7WZy/tO5qFnkaHKDceAR5/zJEkcFM4Jt+7yLrCLsihvTO3BhgDgBdfBxPscS0bHTP6gULGcawh2AHo7/wcascqcVWeMMHCVeEvgqJkF0DkkA07zxGMsQS/FmxXcKVl1H5BK5TeKpTUj/pFULNdNPkkG3tvKUG0OBYKRk6GjSNcKk5XKd+KrewVh2JjwVicolTXjbzY0Y5r6MK6H7KcQaGc1Kr8iFRhXyuZbp8zHvU8mcdzaU18MUKl8CSSBLud9BKr0+rGpiImzUtIMXAZJKVSaxsT+uR8MjqinxKRHOIgeZ1T9DiRZx962dqHgdArbXlKDEF1ekfahtL7j6doj8nujuCZvM+CMxOWXInplsL2LeWEgKB9hVTOOUWkqkhywGteJ24+W4/Kkuxv8BIy2/QuR9LuAlBV5bGpS9fl+T7eIe7Q0gI5D81bsKbCPASSe8SLoewUAIJM/iFyV0BpOcydfa30UzhXrkOGabHeT5PHuL65GxXlXUuPAq9k3NcUoCsZy6pPoWOMaBOmAFS95PEEQkY4RGEyQ5kKU34c1poNyHICF+IcfqKMOpxNQcXgGyMCu3daH5wrCwMD6RiQaOt
    mlflow-password: AgBDORP4PqhDO17fPVeyXA1IH1ExiOoF8x6G4FA34DjwP0yOABwK4wycSpx7koTysbCldLVKta3P1cH6aqFJ7IyQ+bFxRzaoK1kxfCCeakIckTZ9OoDOQAaY71anZYgb9PE54ADXhDxlR4MiTkwRwSwcAYAlERHXj1tjgnNuCz+qc1cg9dHL5zfphNdEgf69fyBq/cGeRn/m/gvr8EdS3u2KwQ98Z4htBQs1Sazew9pJPjPUkPv/CNXoCwH+SX+c3N9qD3vdVoBfUMZCFZI7qRa+NqxQHzXqXw3nNQaW8MBl+enK9NmWJQHShFMxEAMRenfurC65U+goXhPV8zL6iS6Mr2I45MInzeQy8vuUTHT60GQIZ8aILIGF14eVs6hEuMA6GZRppOapQCyzIKpM+3R4m18j+0E0EaHe0/7Hb2nGPqUU0mXPVW3/rckVCChlJNVq2UCpt0BEKHi5bzLubYioHiYWwoHZx83eTKAgORvRJxjR5EUndBLajeXkp8WRgerCwwfvOfZJPcHQhn/kwSsxQYaINegK4aXtIgWeOnyS8Pmh+k0uMOMpmWv1DvbWd1SZl0NAVC3nBYescTRxt1VrSBIWwYbVTmmfTcgXcgrbiPA+uZrncjj7Jhpoogc3TMU1f8Z2lrBngfG8u2zyBmFhX67wTrDJzIZOEyIO/f2dXJaGh4g+wqHrmxRsjYaQABNFunD8h5k=
    postgres-password: AgBdyPSwRDU5IEHinRNvBQaK10adfKPt5xmR2epxYM6w3YU9l2pChIgbwCp4J9tW/2AQkpEgvHM+XWOyv/syPRjoLa11lCcHhKLrB67fEtCb470WxJP09IJTv908K+/qo8RLi0SURvdjN+HRT39P+GI5hZre+36/tqAARXCvv+z5QMfuFIPZaMkz6giVs6+NrmsWhYi6JvIxwL62cujaOS+5wq7VIJ5D049t76DfTdUifeTBJnChEY/IXkc61BRQhwzSrM0JOpzXNh3rc56bR8Sk56ksixbYPXNKA6PKI0H9dCFjdNt5bFSkvIgwIFyPLNMc7655EuQO43TOT/VHvAk6mC7tJfVyHiiWoLNmUFwqaRvZU7mssM69qiz6pC2M/lZ82BVDfa0hfgMitaUPVh297JkEW43iHOBv4QIN6N7nx/xSJ5UttwUf2U4/AoFZJSEDmq8RsaAKqNrCPrHKha55yOlEjMVvGx70V5ybRfzfhqQjiO+kFrnQWJbtHoQLqsdUPmCXC0BHolrqPF9GhL75nN8PGkQEDejsKdYedCyysmNvwDd2tfY5NThEeYjxJKdesg3NfCofWNDWZcDzJBJpU9Vyt2BwjiJhqXXg2RR5iKZwO1lDwisaSqbuDLtiCqurn8PcND6R/Sy4LNGmDrSaEtyjtMmqjvpcjvTj5gPa/4xEY+78JIRs2hSO7lzBA8DNmHFDYr26Ng==
    superset-password: AgB6NIpMPUMgcCUMIK7xgZvE9555xZAr485FYA9TYV77r64F+yIDC2DwhQGEReuhoyBkYHqBJHwGaBrVCoYJ1z/g/evUeG49cl/N5flGF/3YXAbrgP7VxJesxxZHIAQ6cRrRkQgtDtee0rjGFAj20AbhrjKW1BU9n16AxT6xvXIGS2jjwrl5ZG6+ziGH4+wmD2jSoZ11Bru/rmc7gNTWRb2RygyimXx7RdUQKni01A+DVelCaY1AESLeZoJNKGrmXTtH1CNg3+CXxhc/OqEkkeDk14+Bs+Lr+znr736xbwbqyBVTH5Nw7XRu0iK/utquqGL8iUt6yDIcUyyMv2N7TQtRq0R1N+GTXuEezzA5s+k/KYHXI553hoABCEzwbXlgeT03GLcr/cVZ0G0xvdceE1mGE6xDOHQkIQasT0A52LLH5IAHPV30Yhi59NOVAI5UavbgGe/yA5DHYrXl8lxowsD54ctm0QcX5BF0cQA54wPQRVH99M84PDs0trgSpV/vFkWYk7bOLI3cy2S88f4dodXrHFqtnKH3Izf8xj06iu52CEzhvM1Rb2ZeTvkTLK9g5qHgOF1nqdbOX/HgPwmB3cWfQ7IRq7sIKe/XZ3tsABdWnpVUVZ7HLp8UYD8kavhcCxq6gFheUG4yBTEAtX2ZI2c2up3XexLMiCrTEbZ5ORxr8yeQ16+SEh6icsZ5jxBMvGJvf/AwovwDWw==
  template:
    data: null
    metadata:
      creationTimestamp: null
      name: crdb-cockroachdb-init
      namespace: storage
    type: Opaque
```


```(base) [alvaro@stealth cockroachdb]$  git diff templates/
diff --git a/cockroachdb/templates/job.init.yaml b/cockroachdb/templates/job.init.yaml
index 7c7c758..84b9dc2 100644
--- a/cockroachdb/templates/job.init.yaml
+++ b/cockroachdb/templates/job.init.yaml
@@ -146,7 +146,7 @@ spec:
                       {{- range $user := .Values.init.provisioning.users }}
                         CREATE USER IF NOT EXISTS {{ $user.name }} WITH
                         {{- if $user.password }}
-                          PASSWORD '${{ $user.name }}_PASSWORD'
+                          PASSWORD $({{ $user.name }}_PASSWORD)
                         {{- else }}
                           PASSWORD null
                         {{- end }}
diff --git a/cockroachdb/templates/statefulset.yaml b/cockroachdb/templates/statefulset.yaml
index 8572b80..24e5fc4 100644
--- a/cockroachdb/templates/statefulset.yaml
+++ b/cockroachdb/templates/statefulset.yaml
@@ -143,8 +143,8 @@ spec:
         - name: db
           image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
           imagePullPolicy: {{ .Values.image.pullPolicy | quote }}
+          command: [sh, -c]
           args:
-            - shell
             - -ecx
             # The use of qualified `hostname -f` is crucial:
             # Other nodes aren't able to look up the unqualified hostname.
`

Reply:
Tested with provisioned user

```sh
[root@test-release-cockroachdb-0 cockroach]# cockroach sql --certs-dir cockroach-certs/ --user testUser
#
# Welcome to the CockroachDB SQL shell.
# All statements must be terminated by a semicolon.
# To exit, type: \q.
#
Enter password: 
# Server version: CockroachDB CCL v21.1.7 (x86_64-unknown-linux-gnu, built 2021/08/09 17:55:28, go1.15.14) (same version as client)
# Cluster ID: 320ec220-2f65-4b77-a36b-fcf07bf23f88
#
# Enter \? for a brief introduction.
#
testUser@:26257/defaultdb> select 1;
  ?column?
------------
         1
(1 row)

Time: 1ms total (execution 1ms / network 1ms)

testUser@:26257/defaultdb>
```

Reply:
Great @sergeyshaykhullin ! Working flawlessly !